src/schedule_systems.rs

```diff
@@ -85,75 +82,100 @@ pub(crate) fn run_synctest<C: Config>(world: &mut World, mut sess: SyncTestSessi
     // if session is ready, try to advance the frame
-    if sess.current_state() == SessionState::Running {
-        match sess.advance_frame() {
-            Ok(requests) => handle_requests(requests, world),
-            Err(GGRSError::PredictionThreshold) => {
-                info!("P2PSpectatorSession: Waiting for input from host.")
-            }
-            Err(e) => warn!("{e}"),
-        };
-    }
+    let running = sess.current_state() == SessionState::Running;
+    let requests = running.then(|| sess.advance_frame());
 
     world.insert_resource(Session::Spectator(sess));
+
+    match requests {
+        Some(Ok(requests)) => handle_requests(requests, world),
+        Some(Err(GGRSError::PredictionThreshold)) => {
+            info!("P2PSpectatorSession: Waiting for input from host.")
+        }
+        Some(Err(e)) => warn!("{e}"),
+        None => {}
+    };
 }
 
 pub(crate) fn run_p2p<C: Config>(world: &mut World, mut sess: P2PSession<C>) {
-    maybe_init_snapshots(world, sess.max_prediction());
-
     world.insert_resource(LocalPlayers(sess.local_player_handles()));
 
-    if sess.current_state() == SessionState::Running {
+    let running = sess.current_state() == SessionState::Running;
+
+    if running {
         // get local player inputs
         world.run_schedule(ReadInputs);
+
         let local_inputs = world.remove_resource::<LocalInputs<C>>().expect(
             "No local player inputs found. Did you insert systems into the ReadInputs schedule?",
         );
+
         for (handle, input) in local_inputs.0 {
             sess.add_local_input(handle, input)
                 .expect("All handles in local_inputs should be valid");
         }
-
-        match sess.advance_frame() {
-            Ok(requests) => handle_requests(requests, world),
-            Err(GGRSError::PredictionThreshold) => {
-                info!("Skipping a frame: PredictionThreshold.")
-            }
-            Err(e) => warn!("{e}"),
-        };
     }
 
+    let requests = running.then(|| sess.advance_frame());
+
     world.insert_resource(Session::P2P(sess));
+
+    match requests {
+        Some(Ok(requests)) => handle_requests(requests, world),
+        Some(Err(GGRSError::PredictionThreshold)) => {
+            info!("Skipping a frame: PredictionThreshold.")
+        }
+        Some(Err(e)) => warn!("{e}"),
+        None => {}
+    }
 }
 
 pub(crate) fn handle_requests<T: Config>(requests: Vec<GGRSRequest<T>>, world: &mut World) {
     for request in requests {
+        let current_frame = world
+            .get_resource::<RollbackFrameCount>()
+            .map(|frame| frame.0)
+            .unwrap_or_default();
+
+        let confirmed_frame = match world.get_resource::<Session<T>>() {
+            Some(Session::P2P(s)) => Some(s.confirmed_frame()),
+            Some(Session::SyncTest(s)) => {
+                // TODO: `max_prediction` should be replaced with `depth`, but it is private.
```

Comment:
What is depth?

Reply:
Within the `SyncTest` session type there is a parameter, `depth`, which is how far back the rollback will go during testing. So for a `SyncTest`, the confirmed frame is always `frame - depth`.

Reply:
We should just have a getter for that in ggrs.

Reply:
> We should just have a getter for that in ggrs. I'll write it later today :)

Thanks!

Reply:
[It now exists](https://github.com/gschup/ggrs/commit/d062eadaba25ac0b77c85d61ebca032e7c32bc05).